### PR TITLE
Add support for Azure blob storage

### PR DESCRIPTION
**Pull request by PrasadG193**
Signed-off-by: Prasad Ghangal <prasad.ghangal@gmail.com>

### CODE REVIEW

File: repo/blob/azure/azure_storage.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+	// TODO: Handle other possible cases
+	default:
+		return err
+	}
+	return err
+}
+
+func (az *azStorage) PutBlob(ctx context.Context, b blob.ID, data []byte) error {
+	// TODO: Perform chunked upload
```
Comment (vkamra) [inline repo/blob/azure/azure_storage.go]: @PrasadG193 - We may not need to do this since Kopia already breaks data into chunks. @julio-lopez - can you confirm?

File: go.mod [hunk 2]
```diff
@@ -1,10 +1,11 @@
 module github.com/kopia/kopia
 
-go 1.13
+go 1.12
```
Comment (julio-lopez) [inline go.mod]: This should not be changed.
```suggestion
go 1.13
```

File: repo/blob/azure/azure_storage.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+	bucket *gblob.Bucket
+
+	downloadThrottler *iothrottler.IOThrottlerPool
+	uploadThrottler   *iothrottler.IOThrottlerPool
+}
+
+func (az *azStorage) GetBlob(ctx context.Context, b blob.ID, offset, length int64) ([]byte, error) {
+	attempt := func() (interface{}, error) {
+		// TODO: Support chunked download
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: What's needed here is range reads. Is that what you mean?

File: repo/blob/azure/azure_storage.go (hunk 1 above)
Comment (julio-lopez) [reply repo/blob/azure/azure_storage.go]: Chunked uploads are not only not needed, but more importantly not desirable for the Kopia use case.
In order to meet the desired consistency guarantees, `PutBlob` must be done with a single `PUT` operation. Not sure what the semantics are for Azure blob, but other blob stores have weaker consistency guarantees for multi-part uploads.

File: repo/blob/azure/azure_storage.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	attempt := func() (interface{}, error) {
+		return nil, az.bucket.Delete(ctx, az.getObjectNameString(b))
+	}
+
+	_, err := exponentialBackoff(fmt.Sprintf("DeleteBlob(%q)", b), attempt)
+
+	return translateError(err)
+}
+
+// TODO: See if we can use PrefixedBucket instead
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: What does this mean?

File: repo/blob/azure/azure_storage.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	//beforeList := func(as func(interface{}) bool) error {
+	//	// Access storage.Query via q here.
+	//	var q *storage.Query
+	//	if as(&q) {
+	//		_ = q.Delimiter
+	//	}
+	//	return nil
+	//}
+
+	li := az.bucket.List(&gblob.ListOptions{Prefix: az.getObjectNameString(prefix), Delimiter: "/", BeforeList: nil})
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: `Delimiter` should not be used here.
* The blobs with prefix `az.Prefix` are expected to not have '/' in their names after the `az.Prefix` part. Kopia will not directly create blobs with those names.
* Using a delimiter for the listing has slightly different semantics when there are blobs with names of the form `some_prefix<delimiter>...`. All those blobs end up being grouped in a single "metadata" list entry. That's not desirable for this use case.

File: repo/blob/azure/azure_storage.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	"io/ioutil"
+
+	"github.com/efarrer/iothrottler"
+	"github.com/pkg/errors"
+	"gocloud.dev/gcerrors"
+
+	"github.com/kopia/kopia/internal/retry"
+	"github.com/kopia/kopia/repo/blob"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: nit: these imports should go in the 2nd group above.

File: repo/blob/azure/azure_storage.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	gblob "gocloud.dev/blob"
+	"gocloud.dev/blob/azureblob"
+)
+
+const (
+	azStorageType   = "azure"
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: technically, this is _azure blob_, not sure if the type should be `"azureBlob"` or `"azureblob"`

File: repo/blob/azure/azure_storage.go (hunk 3 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Yes. I think better to use NewRangeReader()

File: repo/blob/azure/azure_storage.go (hunk 4 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: We need to check if we can get rid of `getObjectNameString()` if we use PrefixedBucket() function for creating bucket
https://pkg.go.dev/gocloud.dev/blob?tab=doc#PrefixedBucket

File: repo/blob/azure/azure_storage.go (hunk 6 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Sure. I am yet to do cleanup and refactoring

File: repo/blob/azure/azure_storage.go (hunk 7 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Makes sense

File: repo/blob/azure/azure_storage.go (hunk 1 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Ok. Will do the upload in single PUT then

File: repo/blob/azure/azure_storage.go (hunk 4 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Checked. Doesn't work for our use case

**Comment by PrasadG193** [comment]
Test result:

**Comment by julio-lopez** [comment]
Let's create a job to test this before pushing upstream.

**Comment by vkamra** [comment]
@PrasadG193 - regarding @julio-lopez 's comment. We can add an integration test under `go/src/test/integration/kopia` that tests this.

**Comment by vkamra** [comment]
> @PrasadG193 - regarding @julio-lopez 's comment. We can add an integration test under `go/src/test/integration/kopia` that tests this.

Actually - scratch that - we should create a job that runs the kopia tests.

**Comment by julio-lopez** [comment]
> > @PrasadG193 - regarding @julio-lopez 's comment. We can add an integration test under `go/src/test/integration/kopia` that tests this.
> 
> Actually - scratch that - we should create a job that runs the kopia tests.

The job should have no dependencies on K10.

**Comment by PrasadG193** [comment]
Made changes to create azure container if does not exists while executing azure tests

File: cli/storage_azure.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+func init() {
+	var azOptions azure.Options
+
+	RegisterStorageConnectFlags(
+		"azure",
+		"an Azure blob storage",
+		func(cmd *kingpin.CmdClause) {
+			cmd.Flag("container", "Name of the Azure container").Required().StringVar(&azOptions.Container)
+			cmd.Flag("account-name", "Azure account name(overrides AZURE_ACCOUNT_NAME environment variable)").Required().Envar("AZURE_ACCOUNT_NAME").StringVar(&azOptions.AccountName)
```
Comment (vkamra) [inline cli/storage_azure.go]: @PrasadG193  - The default Azure environment variables are `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY`

We should use those names.

See https://docs.microsoft.com/en-us/azure/storage/common/storage-azure-cli

Comment (PrasadG193) [reply cli/storage_azure.go]: @vkamra Looking at gcs tests, https://github.com/kastenhq/kopia/blob/kasten/repo/blob/gcs/gcs_storage_test.go#L21, they are appending `KOPIA_` to the env vars. So I followed the same pattern.

Comment (PrasadG193) [reply cli/storage_azure.go]: Let me know that you think.

Comment (vkamra) [reply cli/storage_azure.go]: I think it's ok to do that in the tests and we can stick with that pattern.

in this package - `cli` - we should use the az cli default environment variables - similar to https://github.com/kastenhq/kopia/blob/kasten/cli/storage_s3.go#L22

Comment (PrasadG193) [reply cli/storage_azure.go]: Fixed!

File: cli/storage_azure.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+func init() {
+	var azOptions azure.Options
+
+	RegisterStorageConnectFlags(
+		"azure",
+		"an Azure blob storage",
+		func(cmd *kingpin.CmdClause) {
+			cmd.Flag("container", "Name of the Azure container").Required().StringVar(&azOptions.Container)
+			cmd.Flag("account-name", "Azure account name(overrides AZURE_STORAGE_ACCOUNT environment variable)").Required().Envar("AZURE_STORAGE_ACCOUNT").StringVar(&azOptions.AccountName)
```
Comment (PrasadG193) [inline cli/storage_azure.go]: Probably, we should also change `--account-name` flag to `--storage-account`?
@vkamra @julio-lopez

File: repo/blob/azure/azure_storage.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+func isRetriableError(err error) bool {
+	if me, ok := err.(azblob.ResponseError); ok {
+		if me.Response() == nil {
+			return true
+		}
+		// retry on server errors, not on client errors
+		return me.Response().StatusCode >= 500
+	}
+
```
Comment (julio-lopez) [inline repo/blob/azure/azure_storage.go]: When would this return anything other than `false` beyond this point?

Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Aah, default case should return `true` in the switch cases

Comment (julio-lopez) [reply repo/blob/azure/azure_storage.go]: Let's add a test for this.

File: repo/blob/azure/azure_storage.go [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+func translateError(err error) error {
+	switch gcerrors.Code(err) {
+	case gcerrors.OK:
+		return nil
+	case gcerrors.NotFound:
+		return blob.ErrBlobNotFound
+	default:
+		return err
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: No. As of now, we have only one error type - `ErrBlobNotFound` in the blob package
https://github.com/kopia/kopia/blob/master/repo/blob/storage.go#L59

File: repo/blob/azure/azure_storage.go (hunk 10 above)
Comment (PrasadG193) [reply repo/blob/azure/azure_storage.go]: Hi julio-lopez,
I have refactored the switch cases for checking error codes. We return `true` only if it is internal server error. I couldn't mock the scenario in the tests because we can only change creds. There is no way to override the default endpoint. I have also added few test negative cases.

**Comment by PrasadG193** [comment]
Tests:
```
$ go test -v ./repo/blob/azure/ -count=1
=== RUN   TestAzureStorage
--- PASS: TestAzureStorage (30.01s)
=== RUN   TestAzureStorageInvalidBlob
--- PASS: TestAzureStorageInvalidBlob (0.52s)
=== RUN   TestAzureStorageInvalidContainer
--- PASS: TestAzureStorageInvalidContainer (0.25s)
=== RUN   TestAzureStorageInvalidCreds
--- PASS: TestAzureStorageInvalidCreds (0.00s)
PASS
ok      github.com/kopia/kopia/repo/blob/azure  30.782s
```

**Comment by julio-lopez** [comment]
superseded by upstream PR.

**Comment by julio-lopez** [comment]
Merged upstream
